src/Nethermind/Nethermind.Optimism/OptimismBlockReceiptTracer.cs

```diff
@@ -30,21 +30,25 @@ public OptimismBlockReceiptTracer(IOPConfigHelper opConfigHelper, IWorldState wo
         if (CurrentTx.IsDeposit())
         {
             depositNonce = _worldState.GetNonce(CurrentTx.SenderAddress!).ToUInt64(null);
+            if (depositNonce > 0)
+            {
+                depositNonce--;
+            }
             if (_opConfigHelper.IsCanyon(header))
             {
                 version = 1;
             }
         }
 
-        return (depositNonce == 0 ? 0 : depositNonce - 1, version);
+        return (depositNonce, version);
     }
 
     protected override TxReceipt BuildReceipt(Address recipient, long spentGas, byte statusCode, LogEntry[] logEntries, Hash256? stateRoot)
     {
         (ulong? depositNonce, ulong? version) = GetDepositReceiptData(Block.Header);
-        TxReceipt receipt = base.BuildReceipt(recipient, spentGas, statusCode, logEntries, stateRoot);
-        receipt.DepositNonce = depositNonce;
-        receipt.DepositReceiptVersion = version;
+        OptimismTxReceipt receipt = new(base.BuildReceipt(recipient, spentGas, statusCode, logEntries, stateRoot))
+        { DepositNonce = depositNonce, DepositReceiptVersion = version };
+
```

Comment:
you are building TxReceipt just to convert it to OptimismTxReceipt, seems wasteful. Better to do a factory method wither in BlockProcessor or ReceiptRootCalculator (a bit not a place for it though) and just fill the properties - base in base method, rest in overriden one.